cpp/src/groupby/sort/group_tdigest.cu

```diff
@@ -0,0 +1,841 @@
+}
+
+// retrieve total weight of scalar inputs by group index
+struct scalar_total_weight {
+  size_type const* group_valid_counts;
+  __device__ double operator()(size_type group_index) { return group_valid_counts[group_index]; }
+};
+
+// return the min/max value of scalar inputs by group index
+template <typename T>
+struct get_scalar_minmax {
+  column_device_view const col;
+  device_span<size_type const> group_offsets;
+  size_type const* group_valid_counts;
+
+  __device__ thrust::tuple<double, double> operator()(size_type group_index)
+  {
+    // note: .element<T>() is taking care of fixed-point conversions for us.
+    return {static_cast<double>(col.element<T>(group_offsets[group_index])),
+            static_cast<double>(
+              col.element<T>(group_offsets[group_index] + (group_valid_counts[group_index] - 1)))};
+  }
+};
+
+struct typed_group_tdigest {
+  template <
+    typename T,
+    typename std::enable_if_t<cudf::is_numeric<T>() || cudf::is_fixed_point<T>()>* = nullptr>
+  std::unique_ptr<column> operator()(column_view const& col,
+                                     cudf::device_span<size_type const> group_offsets,
+                                     cudf::device_span<size_type const> group_labels,
+                                     cudf::device_span<size_type const> group_valid_counts,
+                                     size_type num_groups,
+                                     int delta,
+                                     rmm::cuda_stream_view stream,
+                                     rmm::mr::device_memory_resource* mr)
+  {
+    // first, generate cluster weight information for each input group
+    auto total_weight = cudf::detail::make_counting_transform_iterator(
+      0, scalar_total_weight{group_valid_counts.begin()});
+    auto [group_cluster_wl, group_cluster_offsets, total_clusters] =
+      generate_group_cluster_info(delta,
+                                  num_groups,
+                                  nearest_value_scalar_weights{},
+                                  total_weight,
+                                  cumulative_scalar_weight{group_offsets, group_labels},
+                                  stream,
+                                  mr);
+
+    // device column view. handy because the .element() function
+    // automatically handles fixed-point conversions for us
+    auto d_col = cudf::column_device_view::create(col);
+
+    // compute min and max columns
+    auto min_col = cudf::make_fixed_width_column(
+      data_type{type_id::FLOAT64}, num_groups, mask_state::UNALLOCATED, stream, mr);
+    auto max_col = cudf::make_fixed_width_column(
+      data_type{type_id::FLOAT64}, num_groups, mask_state::UNALLOCATED, stream, mr);
+    thrust::transform(
+      rmm::exec_policy(stream),
+      thrust::make_counting_iterator(0),
+      thrust::make_counting_iterator(0) + num_groups,
+      thrust::make_zip_iterator(thrust::make_tuple(min_col->mutable_view().begin<double>(),
```

Comment:
I hadn't realized that zip iterators could be used to tee output. This is brilliant. Filing this away for future use.

Reply:
The idea came from Jake :)